src/Dscp/DB/SQLite/Queries.hs

```diff
@@ -0,0 +1,313 @@
+{-# LANGUAGE QuasiQuotes #-}
+
+module Dscp.DB.SQLite.Queries where
+
+import Control.Lens (to)
+
+import Database.SQLite.Simple (Only (..), Query)
+import Database.SQLite.Simple.ToRow (ToRow (..))
+
+import Text.InterpolatedString.Perl6 (q)
+
+import Dscp.Core.Serialise ()
+import Dscp.Core.Types (Assignment (..), CourseId, Grade, SignedSubmission (..), StudentId,
+                        Submission (..), aAssignment, aCourseId, aType, sAssignment, sStudentId,
+                        sType, ssSubmission, ssWitness, swSig)
+import Dscp.Crypto (Hash, PublicKey, hash)
+import Dscp.DB.SQLite.Class
+import Dscp.DB.SQLite.Instances ()
+import Dscp.Educator.Txs (PrivateTx (..), PrivateTxId)
+
+data DomainError
+    = CourseDoesNotExist                  CourseId
+    | StudentDoesNotExist                 StudentId
+    | AssignmentDoesNotExist              (Hash Assignment)
+    | StudentWasNotSubscribedOnAssignment StudentId (Hash Assignment)
+    deriving (Show, Typeable)
+
+instance Exception DomainError
+
+type DBM m = MonadSQLiteDB m
+
+getStudentCourses :: DBM m => StudentId -> m [CourseId]
+getStudentCourses student =
+    query getStudentCoursesQuery (Only student)
+  where
+    getStudentCoursesQuery :: Query
+    getStudentCoursesQuery = [q|
+        select  course_id
+        from    StudentCourses
+        where   student_addr = ?
+    |]
+
+enrollStudentToCourse :: DBM m => StudentId -> CourseId -> m ()
+enrollStudentToCourse student course = do
+    execute enrollStudentToCourseRequest (student, course)
+  where
+    enrollStudentToCourseRequest :: Query
+    enrollStudentToCourseRequest = [q|
+        insert into  StudentCourses
+        values       (?, ?)
+    |]
+
+getStudentAssignments :: DBM m => StudentId -> CourseId -> m [Assignment]
+getStudentAssignments student course = do
+    query getStudentAssignmentsQuery (student, course)
+  where
+    getStudentAssignmentsQuery :: Query
+    getStudentAssignmentsQuery = [q|
+        select     course_id, student_addr, desc
+        from       StudentAssignments
+        left join  Assignments
+               on  assignment_hash = Assignments.hash
+        where      student_addr    = ?
+               and course_id       = ?
+    |]
+
+submitAssignment :: DBM m => SignedSubmission -> m (Hash Submission)
+submitAssignment = createSignedSubmission
+
+createSignedSubmission :: DBM m => SignedSubmission -> m (Hash Submission)
+createSignedSubmission sigSub = do
+    let
+        submission     = sigSub^.ssSubmission
+        submissionSig  = sigSub^.ssWitness.swSig
+
+        student        = submission^.sStudentId
+        submissionHash = submission^.to hash
+        submissionType = submission^.sType
+        assignment     = submission^.sAssignment
+
+        course         = assignment^.aCourseId
+        assignmentHash = assignment^.to hash
+
+    transaction $ do
```

Comment:
@Heimdell This question yet remains, will `transaction` function from `sqlite-simple` provide enough isolation level to prevent race conditions?

Reply:
To get a race condition:
1) We either `fork` before we open `RIO` context - in this case its not a problem, because we have 2 connections here.
2) We `fork` inside `RIO` - in that case, I propose implementing [MonadFork](https://hackage.haskell.org/package/monad-fork-0.1/docs/Control-Monad-Fork-Class.html) that way so it duplicates the connection (and maybe other resources - logging might become interleaved, too). To do that, we might store db name along with connection. This means, we also have to prohibit somehow - by conventions - any other fork methods except `fork`.

Reply:
One of the place there we would have forks - requests to servant api, they happen each in its own thread in potentially arbitrary amounts. 

So we need
1. to have a limited pool of connections
2. In servant fork happens in IO, instead you control how `m a -> IO a` tranformation occurs, but clonning db connection there looks like a bad practice since it's not prohibited for servant to perform tranformation more one once per every endpoint call (although this is arguable)

Both points are better achievable at DB level, not at fork level.